hydroflow/src/compiled/pull.rs

```diff
@@ -97,9 +121,83 @@ where
     }
 }
 
+pub struct CrossJoin<I1, V1, I2, V2>
+where
+    V1: Eq + Clone,
+    V2: Eq + Clone,
+    I1: Iterator<Item = V1>,
+    I2: Iterator<Item = V2>,
+{
+    lhs: I1,
+    rhs: I2,
+    state: CrossJoinState<V1, V2>,
+}
+
+impl<I1, V1: 'static, I2, V2: 'static> Iterator for CrossJoin<I1, V1, I2, V2>
+where
+    V1: Eq + Clone,
+    V2: Eq + Clone,
+    I1: Iterator<Item = V1>,
+    I2: Iterator<Item = V2>,
+{
+    type Item = (V1, V2);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            match self.state.eventside {
+                Side::Left => {
+                    if let Some(i) = self.state.opposite_ix.next() {
+                        // TODO(justin): unnecessary clone (sometimes).
```

Comment:
I don't think the clone is unnecessary here since we're guaranteed to have at least one match here?

Reply:
correct